### DISCUSSION TITLE
Question about secure tracing.

### DISCUSSION THREAD

**Discussion by jzeng4** [aquasecurity/tracee#445] [Development]
After reading the "secure tracing" section, I don't understand the followings:

"When Tracee reads information from user programs it is subject to a race condition where the user program might be able to change the arguments after Tracee has read them."

If I understand correctly, Tracee reads information (syscall arguments) in syscall entry. How is it possible for the user program to change the arguments after syscall has been invoked? The program ^^ is malware or vulnerable software? Could you provide more details or any examples about it? Thank you!

**Comment by yanivagman** [comment]
Think threads and pointers.
There are syscalls (e.g. open, execve) which use pointers (e.g. for string values) as their arguments. These point to a memory location in the process address space, which can be changed by any thread of the process. This may lead to a race condition, where one thread calls a syscall with one argument, and the other thread immediately change the argument value. A malicious software can try to exploit this race condition in what is called a TOCTTOU attack.
Also have a look at https://lwn.net/Articles/245630/
